### PR TITLE
Fuel merge part 7 - Finish codegen of public API and improve decoding perf

### PR DESCRIPTION
**Pull request by DZakh**
- Codegen rescript schema for Fuel types
- Codegen mock function for Fuel
- Optimise items decoding for all ecosystems

### CODE REVIEW

File: codegenerator/cli/templates/static/codegen/src/eventFetching/chainWorkers/HyperSyncWorker.res
Comment (DZakh) [inline codegenerator/cli/templates/static/codegen/src/eventFetching/chainWorkers/HyperSyncWorker.res:306]: Removed object allocation per item

File: codegenerator/cli/templates/static/codegen/src/eventFetching/chainWorkers/HyperSyncWorker.res
Comment (DZakh) [inline codegenerator/cli/templates/static/codegen/src/eventFetching/chainWorkers/HyperSyncWorker.res:314]: Improved error message (I'm not sure about null, but I don't have an exception here)

File: codegenerator/cli/templates/static/codegen/src/ContractInterfaceManager.res
Comment (DZakh) [inline codegenerator/cli/templates/static/codegen/src/ContractInterfaceManager.res:190]: Removed 2 object allocations per item

File: codegenerator/cli/src/fuel/abi.rs
Comment (JonoPrest) [inline codegenerator/cli/src/fuel/abi.rs:104]: Is it needed for any cases or just trying to be idiomatic with rescript?

2 Reasons I think we shouldn't uncapitalized
1. `a` and `A` would be valid param names treated as separate parameters (we would be collapsing this case)
2. Typescript generics are typically capitalised so we would be trading off idiomatic TS for idiomatic rescript

File: codegenerator/cli/src/rescript_types.rs
Comment (JonoPrest) [inline codegenerator/cli/src/rescript_types.rs:47]: Nice, I can think of a few use cases for sorting deps like this. The alternative is just to use a `rec` definition with `and` between each one then you don't have to care about ordering.

File: codegenerator/cli/templates/static/codegen/src/EventFetching.res
Comment (JonoPrest) [inline codegenerator/cli/templates/static/codegen/src/EventFetching.res:146]: Why do you prefer raw null over a matchable/searchable exception?

File: codegenerator/cli/templates/static/codegen/src/eventFetching/chainWorkers/HyperSyncWorker.res
Comment (JonoPrest) [inline codegenerator/cli/templates/static/codegen/src/eventFetching/chainWorkers/HyperSyncWorker.res:326]: I think we should explore our own "context" modelling soon. For instance I think that chainId probably already exists on the parent logger but have no idea and deduplication isn't handled.

File: codegenerator/cli/templates/static/codegen/src/eventFetching/chainWorkers/HyperSyncWorker.res
Comment (JonoPrest) [reply codegenerator/cli/templates/static/codegen/src/eventFetching/chainWorkers/HyperSyncWorker.res:334]: I would suggest in these cases just create a new exception type that's searchable and could be pattern matched by rescript. But Either way think we can improve our error handling model

File: codegenerator/cli/templates/static/codegen/src/eventFetching/hyperfuel/HyperFuel.res
Comment (JonoPrest) [inline codegenerator/cli/templates/static/codegen/src/eventFetching/hyperfuel/HyperFuel.res:207]:
```suggestion
```

**Review by JonoPrest — APPROVED** [review]
Hey Dmitry, definitely keen to get this in sooner rather than later, there will be lots of conflicts for me to deal with but could you just get back to my comments before merging?

File: codegenerator/cli/src/fuel/abi.rs
Comment (DZakh) [reply codegenerator/cli/src/fuel/abi.rs:104]: When we try to get a schema for the generic like `<genericName>Schema` it fails because of invalid var name. Maybe we can prefix it for the case to be `generic<genericName>Schema`

Comment (DZakh) [reply codegenerator/cli/src/fuel/abi.rs:104]: This logic is not bullet proof at all. I just made it compile for the greeter template I use for tests

File: codegenerator/cli/src/rescript_types.rs
Comment (DZakh) [reply codegenerator/cli/src/rescript_types.rs:47]: > Nice, I can think of a few use cases for sorting deps like this. The alternative is just to use a rec definition with and between each one then you don't have to care about ordering.

It only works when the value is a function. I tried it first, but didn't work.

> Actually with that in mind, can there ever be a cyclical/recursive dependency?

Thankfully, they are not supported in Fuel. Otherwise the code didn't work

File: codegenerator/cli/templates/static/codegen/src/EventFetching.res
Comment (DZakh) [reply codegenerator/cli/templates/static/codegen/src/EventFetching.res:146]: Should I create a dummy exception for this case?

File: codegenerator/cli/templates/static/codegen/src/eventFetching/chainWorkers/HyperSyncWorker.res
Comment (DZakh) [reply codegenerator/cli/templates/static/codegen/src/eventFetching/chainWorkers/HyperSyncWorker.res:326]: Yes, and make it lazily computed (so the object is allocated only when an error is thrown)

File: codegenerator/cli/src/fuel/abi.rs
Comment (JonoPrest) [reply codegenerator/cli/src/fuel/abi.rs:104]: Ok cool, I'll leave it to you 👍🏼

File: codegenerator/cli/src/rescript_types.rs
Comment (JonoPrest) [reply codegenerator/cli/src/rescript_types.rs:47]: Sweet 👍🏼

File: codegenerator/cli/templates/static/codegen/src/EventFetching.res
Comment (JonoPrest) [reply codegenerator/cli/templates/static/codegen/src/EventFetching.res:146]: I don't really mind but I would have probably defined a unique rescript exception and passed it in. Or perhaps add an API to errorHandling that doesn't take an exception but just throws the message. Either way the abstraction is obviously not fitting here.

**Pull request merged**
